getodk/collect
# Allow user to recover from multiple forms with same formid/version and prevent their future download

Closes #4237

Strongly recommend reviewing commit-by-commit.

If this were not a legacy project we'd want to use a hash to represent a unique form. An improvement we could make is linking instances and the forms they were filled from through database ids rather than through formid/version. However, that's an invasive change that I don't think we should take on so close to a planned release. It's probably not reasonable to take on at all before some of the mission-critical parts of the code have better test coverage.

#### What has been done to verify that this works as intended?
Added automated tests for the changes to form downloads and delete behavior. I don't think there are low level tests to write for the changes to instance load and finalizing given the current code structure and I didn't want to get into a risky refactor. I think the problem is enough of an edge-case that I don't want to introduce Espresso tests. I verified those manually for now.

Cases
- adb push form with same formid/version and same filename. Collect should replace no matter the contents. Can get in a bad state if you have filled forms for one older version and there are schema changes. Won't be able to open the filled forms.
- adb push two identical forms (same md5 hash) but different filenames. Collect will treat these as different and this will cause the bad state from the issue.
- adb push same formid/version but the form contents are different (different md5 hashes). Collect will treat these as different and this will cause the bad state from the issue.
- try to download form with same formid/version and same filename. Collect should download and replace. There may or may not be new media attachments.
- try to download form with same formid/version/contents but different filenames. Collect should download but treat as different form. This will cause the bad state from the issue.
- try to download form with same formid/version/filename but different contents. Collect should prevent download.
- get a form on the device. Fill it out a few times. Delete it. Get a form with the same
- got into a bad state in v1.28 through one of the ways above, soft deleted one of the forms but was still suck
- got into a bad state in v1.28 through one of the ways above, had not yet deleted one of the forms

#### Why is this the best possible solution? Were any other approaches considered?
I think this is a good balance of not too invasive and pretty thorough. It both allows folks who ended up in a bad state in v1.28 to recover and limits the ways that the bad state can be reached in the future.

The first two commits filter out soft deleted forms when finalizing or opening filled forms. After those, it should be possible to recover from the bad state by deleting one of the blank forms. This helps folks who already ended up with the problem in v1.28 and soft-deleted one of the copies as instructed.

The third commit prevents downloading a form with the same formid/version as a form already on the device but with different contents. It still allows downloading a form with the same formid/version and the same contents because some servers will allow that to be combined with a media file update (Aggregate but not Central). This should make it harder for users to get into the bad state. There is still the possibility that the remote form would have the same contents but a different filename. We can't guard against that before downloading because the form XML endpoint may not include the filename. We'd have to do the check after downloading the file and I don't think this is common enough to add a special case. Plus, the two first commits will allow users to recover. **Note**: there should be a helpful message from manual download or automatic form update. However, in the match exactly form update mode, there will just be a generic "Failure" message, including if the refresh button is tapped manually.

#### How does this change affect users? Describe intentional changes to behavior and behavior that could have accidentally been affected by code changes. In other words, what are the regression risks?
This affects form load, form finalization, form download and form delete. There are opportunities for regression in any of these but I tried to make the changes easy to review to mitigate that. The challenge is that there are a lot of possible combinations to think of. I tried to list those above in the verification section.

I think there's also a fair amount of risk around the scenarios that we added soft delete for. That is, after deleting a form, its instances should still be editable and finalizable. I tried to verify all that manually and also `DeleteBlankFormTest` was useful.

#### Do we need any specific form for testing your changes? If so, please attach one.
No, any form will do.

#### Does this change require updates to documentation? If so, please file an issue [here]( https://github.com/getodk/docs/issues/new) and include the link below.
I don't think so. It doesn't seem common enough to mention.

#### Before submitting this PR, please make sure you have:
- [x] run `./gradlew checkAll` and confirmed all checks still pass OR confirm CircleCI build passes and run `./gradlew connectedDebugAndroidTest` locally.
- [x] verified that any code or assets from external sources are properly credited in comments and/or in the [about file](https://github.com/getodk/collect/blob/master/collect_app/src/main/assets/open_source_licenses.html).
- [x] verified that any new UI elements use theme colors. [UI Components Style guidelines](https://github.com/getodk/collect/blob/master/CONTRIBUTING.md#ui-components-style-guidelines)
